### PR TITLE
Add an Antora-based site generator

### PR DESCRIPTION
**Pull request by Mogztter**

### CODE REVIEW

**Comment by mraible** [comment]
Does this retain existing functionality where Asciidoctor produces PDF and ePUB files?

**Comment by Mogztter** [comment]
@mraible I just noticed that PDF produced with the latest version of Asciidoctor PDF has a few issues... I will investigate 👀

**Comment by Mogztter** [comment]
@mraible I think it would be great if the EPUB + PDF files were downloadable from the Antora site?

image

What do you think?

**Comment by Mogztter** [comment]
Preview: https://6065b6fd706631000824d45c--zen-lamport-50117f.netlify.app/

**Comment by mraible** [comment]
@Mogztter I like the idea! However, it'd be nice if it were a configuration option. For someone like InfoQ, they might still want the downloads to still happen from a dedicated page. Also, I don't think it's necessary to give an option to download the print PDF since that's only meant for printing a hard-copy book. 

I checked the preview site and I don't see the download options.

**Comment by Mogztter** [comment]
> I checked the preview site and I don't see the download options.

Sorry I wasn't clear, it was just a proposal/mockup! 😉 

> However, it'd be nice if it were a configuration option. For someone like InfoQ, they might still want the downloads to still happen from a dedicated page.

That's a good point.
We can definitely make it configurable.

> Also, I don't think it's necessary to give an option to download the print PDF since that's only meant for printing a hard-copy book.

OK 👍🏻

**Comment by mraible** [comment]
Thinking about this a bit more, I think it'd be cool if the download URL could be configurable. That way, you could redirect the user to a signup page for downloading the artifacts. Ideally, they could login and be redirected back to get the files, but something like that seems difficult for a static site.

**Comment by Mogztter** [comment]
> Thinking about this a bit more, I think it'd be cool if the download URL could be configurable. That way, you could redirect the user to a signup page for downloading the artifacts. 

👍🏻 

> Ideally, they could login and be redirected back to get the files, but something like that seems difficult for a static site.

In theory it would work but it would probably require some JavaScript to check if the user is authenticated or not.

Having said that we can add this feature later. Should we proceed and merge this pull request? Do you want to host/publish the site on Netlify? or elsewhere?

File: package.json [hunk 1]
```diff
@@ -11,10 +11,15 @@
...
+    "build:site": "antora --generator=mini-book-site-generator --html-url-extension-style=indexify site.yml",
+    "watch:site": "node watch-site.js"
   },
   "devDependencies": {
+    "@antora/cli": "^2.3.4",
     "browser-sync": "^2.26.14",
-    "gulp": "4.0.2"
+    "chokidar": "^3.5.1",
+    "gulp": "4.0.2",
+    "mini-book-site-generator": "git+https://github.com/Mogztter/mini-book-site-generator.git#main"
```
Comment (mraible) [inline package.json:23]: Will this eventually point to a released dependency?

File: site.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  title: InfoQ Mini Book
+content:
+  sources:
+    - url: '.'
+      branches: HEAD
+      start_path: src/docs/asciidoc
+ui:
+  bundle:
+    url: https://github.com/Mogztter/mini-book-default-ui/releases/download/v0.2.1/ui-bundle.zip
+    snapshot: true
```
Comment (mraible) [inline site.yml:13]: I'm guessing this will eventually change to a released version?

File: src/docs/asciidoc/antora.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+name: ROOT
+title: The JHipster Mini-Book
+version: master
```
Comment (mraible) [inline src/docs/asciidoc/antora.yml:3]: Change we change this to main if it refers to a GitHub branch? I no longer use master as a branch name.

File: src/docs/asciidoc/chapters/chapter3.adoc [hunk 4]
```diff
@@ -471,8 +471,8 @@ Bootstrap also includes a number of utility classes that can be used to show and
 
 Bootstrap's classes for setting the display are names using the following format:
 
-* `.d-{value}` for `xs`
-* `.d-{breakpoint}-{value}` for `sm`, `md`, `lg`, and `xl`
+* `.d-\{value}` for `xs`
+* `.d-\{breakpoint}-\{value}` for `sm`, `md`, `lg`, and `xl`
```
Comment (mraible) [inline src/docs/asciidoc/chapters/chapter3.adoc:475]: Do I always have to escape curly braces from now on? I'd prefer it if the AsciiDoc I write doesn't have to change.

File: src/docs/asciidoc/chapters/dedication.adoc [hunk 5]
```diff
@@ -1,4 +1,5 @@
 = Dedication
+:page-pagination: next
```
Comment (mraible) [inline src/docs/asciidoc/chapters/dedication.adoc:2]: Why is this necessary?

File: watch-site.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
+const chokidar = require('chokidar')
+const bs = require("browser-sync").create()
+const generator = require('mini-book-site-generator')
```
Comment (mraible) [inline watch-site.js:3]: I'm not sure that 'mini-book' is the appropriate prefix for this project. I'd like to think it can be for anyone writing Asciidoctor-based books. `asciidoctor-book-site-generator` might be better. Maybe "antora" would be OK, but I tend to think that asciidoctor is still the thing most people think of.

File: watch-site.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    /(^|[\/\\])\../, // ignore dotfiles
+  ],
+  persistent: true,
+  // avoid issues with Intellij IDEA temporary files (used to automatically save changes)
+  // [Error: ENOENT: no such file or directory, lstat '/path/to/.../file.adoc~' in /path/to/... (ref: antora-local-dev <worktree> | path: modules/...)] {
+  //   errno: -2,
+  //   code: 'ENOENT',
+  //   syscall: 'lstat',
+  //   path: '/path/to/.../file.adoc~'
+  // }
```
Comment (mraible) [inline watch-site.js:22]: Can this section be deleted?

**Review by mraible — COMMENTED** [review]
Added some comments with questions.

**Comment by mraible** [comment]
> Having said that we can add this feature later. Should we proceed and merge this pull request? Do you want to host/publish the site on Netlify? or elsewhere?

That's probably a good idea. No reason to pack too many features into the first release!

We can merge this pull request, but I don't want to publish anything quite yet. I still need to talk to InfoQ and see what their stance on this is. They've allowed me to open source the book, but I haven't advertised it, so no one really knows about it. 

I was hoping to update this book to JHipster v7 before doing that. I'm not sure if InfoQ will want an HTML version published. If they're OK with it, I'm not sure if they'd want it on their site or if they'd allow me to publish on my own.

I'm working on another mini-book with them. I'm six months past my deadline on that one, but they agreed to let me open source it after it's finished and on their site for at least three months. 

I'll talk to them and get back to you.

File: package.json (hunk 1 above)
Comment (Mogztter) [reply package.json:23]: Yes, I just released version 0.1.0!

File: site.yml (hunk 2 above)
Comment (Mogztter) [reply site.yml:13]: This is a released version, the artifact (zip) is built from a tag/release using GitHub Actions: https://github.com/Mogztter/mini-book-default-ui/blob/master/.github/workflows/release.yml

File: src/docs/asciidoc/antora.yml (hunk 3 above)
Comment (Mogztter) [reply src/docs/asciidoc/antora.yml:3]: Nice catch, you're correct I will update the branch name

File: src/docs/asciidoc/chapters/chapter3.adoc (hunk 4 above)
Comment (Mogztter) [reply src/docs/asciidoc/chapters/chapter3.adoc:475]: It was already an issue but a silent one. Antora is using `attribute-missing=warn`, that's how I found them.

In this case Asciidoctor will try to resolve `{breakpoint}` and `{value}`, fortunately you didn't define these attributes so they won't be replaced... but it's "unsafe".
If you don't like this syntax, we can alternatively use:

```
`+.d-{breakpoint}-{value}+`
```

File: src/docs/asciidoc/chapters/dedication.adoc (hunk 5 above)
Comment (Mogztter) [reply src/docs/asciidoc/chapters/dedication.adoc:2]: It's not and we can remove it.
The idea was to showcase how we can control the pagination to remove/hide the previous and/or next link.
In this case, only the next link will be shown.

Do you want to remove it?

File: watch-site.js (hunk 7 above)
Comment (Mogztter) [reply watch-site.js:22]: It's a note for our future selves to explain why we need to add/configure `awaitWriteFinish`. In my opinion, we should keep it but if you insist, I can remove it.

File: watch-site.js (hunk 6 above)
Comment (Mogztter) [reply watch-site.js:3]: > I'm not sure that 'mini-book' is the appropriate prefix for this project. I'd like to think it can be for anyone writing Asciidoctor-based books.

In a way that's true but it was still designed to produce mini-book site.
In practice, anyone writing AsciiDoc-based books could use it but they would probably have new requirements that are not supported/implemented.
So I don't want to promote it as a general purpose generator as it might be a little deceptive.

> asciidoctor-book-site-generator might be better. Maybe "antora" would be OK, but I tend to think that asciidoctor is still the thing most people think of.

Both Antora and Asciidoctor are implementation details. Even if we heavily rely on Antora, I don't think it adds value to add antora in the name.
We shouldn't use Asciidoctor because Antora will possibly use another AsciiDoc JavaScript implementation in the future... in fact, it's almost certain!

AsciiDoc is now an Eclipse Trademark so we cannot use asciidoc either.

**Comment by Mogztter** [comment]
Thanks for your review.

> We can merge this pull request, but I don't want to publish anything quite yet. I still need to talk to InfoQ and see what their stance on this is. They've allowed me to open source the book, but I haven't advertised it, so no one really knows about it.

Ok 👍🏻 
Once approved, I will go ahead and merge it. Since Netlify is not configured on this repository the site won't be published.
I will also unpublish/delete my preview site: https://6065b6fd706631000824d45c--zen-lamport-50117f.netlify.app/

> I was hoping to update this book to JHipster v7 before doing that. I'm not sure if InfoQ will want an HTML version published. If they're OK with it, I'm not sure if they'd want it on their site or if they'd allow me to publish on my own.

👌🏻 

> I'm working on another mini-book with them. I'm six months past my deadline on that one, but they agreed to let me open source it after it's finished and on their site for at least three months.

I wish you good luck!

> I'll talk to them and get back to you.

Perfect!

**Comment by mraible** [comment]
Hello @Mogztter. Is it possible to re-publish the mini-site or provide instructions for me to do it? It'd probably be good to document how to do it in the README anyway?

The reason I ask is because I haven't heard back from InfoQ and just realized the link I sent them no longer works. I want to forward my email to other folks, but it's difficult to explain without the visual asset.

Thanks!

Matt

**Comment by Mogztter** [comment]
Hi @mraible 

> Is it possible to re-publish the mini-site or provide instructions for me to do it? It'd probably be good to document how to do it in the README anyway?

Sure!
If you have a Netlify account, it's straightforward:

1. Click on "New site from Git"
2. Click on "GitHub"
 
image

3. Select `mraible/jhipster-book` repository (if you don't see your repo, click on "Configure the Netlify app on GitHub." and then select it)
4. Configure site settings
   * In "Branch to deploy", select "site-generator"
   * In "Build command", set `npm run build:site`
   * In "Publish directory", set "public/"

image

5. Click "Deploy site"
6. Wait 1 minute
7. Click on the link provided by Netlify

I just tested this procedure and here's the result: https://practical-goldstine-f236b1.netlify.app/

Please note that in practice, we would probably use a [file-based configuration](https://docs.netlify.com/configure-builds/file-based-configuration/) but we can do that once everything has been settled with InfoQ.

**Comment by mraible** [comment]
@Mogztter If I look at this branch, there already appears to be a `netlify.toml` file. As far as I know, that is the file-based deployment. Is there anything else we need to do for this (besides document it in the README)?

**Comment by Mogztter** [comment]
> If I look at this branch, there already appears to be a netlify.toml file. As far as I know, that is the file-based deployment. 

You're right I forgot about it.

> Is there anything else we need to do for this (besides document it in the README)?

No, it should be all good, once this pull request is merged, I believe that Netlify will use this `netlify.toml` file to build and deploy the site.

**Comment by mraible** [comment]
@Mogztter Is it possible to change this part at the bottom of every page? If so, where?

image

**Comment by Mogztter** [comment]
> Is it possible to change this part at the bottom of every page? If so, where?

Yes, we can either update it in the mini book UI: https://github.com/Mogztter/mini-book-default-ui/blob/master/src/partials/footer-content.hbs (global) or override it using `supplemental_files` in this repository (local).

@mraible What do you prefer?

**Comment by Mogztter** [comment]
Here's an example on how to override the footer using the supplemental files feature: [`c454a77` (#12)](https://github.com/mraible/jhipster-book/pull/12/commits/c454a77559057947c2ca417ebaa0a6c05bcca0ab)

**Comment by mraible** [comment]
@Mogztter I'd like to merge this, but I'm also interested in integrating it into https://github.com/mraible/infoq-mini-book and https://github.com/mraible/angular-book (currently private). Is this something you'd be interested in doing? If so, can you provide a cost estimate?

**Comment by Mogztter** [comment]
@mraible Sure, I will take a closer look and let you know. I assume that `mraible/angular-book` is based on the `mraible/infoq-mini-book` template?

For reference, Antora 3 is just around the corner and will provide a lightweight, event-based extension facility: https://docs.antora.org/antora/3.0/whats-new/#antora-extensions
With this new feature, we will be able to greatly simplify the generator: https://github.com/mogztter/mini-book-site-generator

**Comment by mraible** [comment]
> Sure, I will take a closer look and let you know. I assume that mraible/angular-book is based on the mraible/infoq-mini-book template?

Yep! I'll give you access to the repo. 

That's great to hear about Antora 3! I noticed that your mini-book-site-generator repo doesn't have a README. I'd be happy to pay you to fill that out so others can integrate things easier.

**Comment by Mogztter** [comment]
> Yep! I'll give you access to the repo.

I now have access, thanks!

> That's great to hear about Antora 3! I noticed that your mini-book-site-generator repo doesn't have a README. I'd be happy to pay you to fill that out so others can integrate things easier.

I can explain what the generator does/provides but I don't want to get into much details.
The reason is that, currently, most "packages" are duplicated because Antora 2 does not provide a mechanism to override or extend the default generator. 
Once the migration to Antora 3 is complete, I will add a section to explain how to contribute.

> I also think it'd be a good idea to list the releases on GitHub. I use the following GitHub action to auto-publish to npmjs.com when I create new releases.
> oktadev/schematics@main/.github/workflows/publish.yml

Yes, that's a good idea, I'm using something similar on other projects as well 👍🏻

File: src/docs/asciidoc/antora.yml (hunk 3 above)
Comment (Mogztter) [reply src/docs/asciidoc/antora.yml:8]: I've restored the value `master` and explain why in the `antora.yml` file.
